python/cudf/cudf/core/dataframe.py

```diff
@@ -1321,10 +1321,16 @@ def assign(self, **kwargs):
         1  1  4
         2  2  5
         """
-        new = self.copy()
+        new_df = cudf.DataFrame(index=self.index.copy())
+        for name, col in self._data.items():
+            if name in kwargs:
+                new_df[name] = kwargs.pop(name)
+            else:
+                new_df._data[name] = col.copy()
+
         for k, v in kwargs.items():
-            new[k] = v
-        return new
+            new_df[k] = v
+        return new_df
```

Comment:
Can we build up a dict instead of directly building up a DataFrame? That avoids calling the DataFrame constructor like we're doing everywhere else in this PR. We can also avoid all the expensive `__setitem__` calls.

Reply:
We actually need to go through `__setitem__` calls for this user-facing API unfortunately 😞 . The reason is we want the scalar handling, index handling, etc. to match what pandas support.